karapace/schema_reader.py

```diff
@@ -492,52 +466,22 @@ def _handle_msg_schema(self, key: dict, value: Optional[dict]) -> None:
         # won't interfere with the equality. Note: This means it is possible
         # for the REST API to return data that is formatted differently from
         # what is available in the topic.
-        if schema_type_parsed in [SchemaType.AVRO, SchemaType.JSONSCHEMA]:
-            try:
-                schema_str = json.dumps(json.loads(schema_str), sort_keys=True)
-            except json.JSONDecodeError:
-                LOG.error("Schema is not valid JSON")
-                return
-        elif schema_type_parsed == SchemaType.PROTOBUF:
-            try:
-                parsed_schema = ValidatedTypedSchema.parse(SchemaType.PROTOBUF, schema_str)
-                schema_str = str(parsed_schema)
-            except InvalidSchema:
-                LOG.exception("Schema is not valid ProtoBuf definition")
-                return
-
-        if schema_subject not in self.subjects:
-            LOG.info("Adding first version of subject: %r with no schemas", schema_subject)
-            self.subjects[schema_subject] = {"schemas": {}}
-
-        subjects_schemas = self.subjects[schema_subject]["schemas"]
-
-        with self.id_lock:
-            # dedup schemas to reduce memory pressure
-            schema_str = self._hash_to_schema.setdefault(hash(schema_str), schema_str)
-
-            if schema_version in subjects_schemas:
-                LOG.info("Updating entry subject: %r version: %r id: %r", schema_subject, schema_version, schema_id)
-            else:
-                LOG.info("Adding entry subject: %r version: %r id: %r", schema_subject, schema_version, schema_id)
+        try:
+            schema_str = TypedSchema.normalize_schema_str(schema_str, schema_type_parsed)
+        except (json.JSONDecodeError, InvalidSchema):
+            return
```

Comment:
issue: It looks like we're doing double validation here, as instantiating `TypedSchema` below also runs this validation. Should we skip this one, and cover the instantiation with an equivalent try-except instead?

Reply:
Done.